src/camera/projection_math.jl

```diff
@@ -22,29 +22,29 @@ function translationmatrix(t::Vec{3, T}) where T
     )
 end
 
-rotate(angle::T, axis::Vec{3, T}) where {T} = rotationmatrix4(qrotation(convert(Array, axis), angle))
+rotate(angle, axis::Vec{3}) = rotationmatrix4(qrotation(convert(Array, axis), angle))
 rotate(::Type{T}, angle::Number, axis::Vec{3}) where {T} = rotate(T(angle), convert(Vec{3, T}, axis))
 
-function rotationmatrix_x(angle::T) where T
-    T0, T1 = zero(T), one(T)
+function rotationmatrix_x(angle)
+    T0, T1 = (0, 1)
```

Comment:
can we leave this? It's important to keep the type!

Reply:
We certainly don't have to change it. But I am curious why not. The matrix will have a uniform type. The function as written will break, for example, for `rotationmatrix_x(1)`, since `sin(1)::Float64`. It will also break for `UnitfulAngles` numbers.

Reply:
Don't you lose some generality by using 1 and 0 instead of one(T) and zero(T)?

Reply:
I don't think so, not in any important case. Though if someone can give a more satisfying answer, that would be better.


```
function all_subtypes(T)
  Ts= Set([T])
  while true
    n_before = length(Ts)
    more = Set(vcat(subtypes.(Ts)...))
    Ts = union(Ts, more)
    n_after = length(Ts)
    if n_after == n_before
      break
    end
  end
  return Ts
end

Numbers = collect(all_subtypes(Number))
Numbers = filter(!isabstracttype, Numbers)

[(old_type, promote_type(Int, old_type)) for old_type in Numbers if promote_type(Int, old_type) != old_type]
```

```
 (Bool, Int64)
 (UInt16, Int64)
 (Irrational, Float64)
 (Int8, Int64)
 (Unitful.Gain{L,S,T} where T<:Real where S where L<:Unitful.LogInfo, Number)
 (UInt32, Int64)
 (Unitful.Level{L,S,T} where T<:Number where S where L<:Unitful.LogInfo, Number)
 (UInt8, Int64)
 (Rational, Real)
 (Unitful.Quantity, Number)
 (Int16, Int64)
 (Complex, Number)
 (Int32, Int64)
```

Reply:
Ah you're right.
This is the case I'm worried about:
```julia
angle = 1f0
T0, T1 = (0, 1)

x = Mat{4}(
           cos(angle), T0, -sin(angle),  T0,
           T0, T1, T0, T0,
           sin(angle), T0, cos(angle), T0,
           T0, T0, T0, T1
       )
@assert x isa Matf4f0
```
Which works fine with your changes :) I guess, Julia 1.0 changed the promotion rules!

Reply:
I can add this as a test, too, in case Julia ever changes it back! Where should it live? (Is this function used by `Makie`?)

Reply:
That test would probably be better in `AbstractPlotting`'s tests, feel free to make a new file and testset :)

Reply:
It may be slightly cleaner to do `T0, T1 = (false, true)`  (which is the eltype of `I`) so that you are sure that the promotion doesn't increase the precision. Though in practice it would only matter if `cos(angle)` was for example `Int32` but that shouldn't happen.

Reply:
I think you mean `im` (`Complex(false, true)`)? I think you are technically right, though I prefer 0 and 1 aesthetically. I'll let someone else make the final call.